python/ray/data/_internal/block_batching.py

```diff
@@ -119,74 +99,222 @@ def get_batches(block: Optional[ObjectRef[Block]] = None) -> Iterator[BatchType]
+        except StopIteration:
+            pass
+        yield block_ref
         if clear_block_after_read:
             ray._private.internal_api.free(block_ref, local_only=False)
-        window.append(elem)
-        yield tuple(window)
+
+
+def _blocks_to_batches(
+    block_iter: Iterator[Block],
+    stats: Optional[Union[DatasetStats, DatasetPipelineStats]] = None,
+    batch_size: Optional[int] = None,
+    drop_last: bool = False,
+    shuffle_buffer_min_size: Optional[int] = None,
+    shuffle_seed: Optional[int] = None,
+    ensure_copy: bool = False,
+) -> Iterator[Block]:
+    """Given an iterator over blocks, returns an iterator over blocks
+    of the appropriate bacth size.
+
+    If the shuffling configurations are specified, then the
+    output blocks contain shuffled data.
+
+    Args:
+        block_iter: An iterator over blocks.
+        stats: Dataset stats object used to store block batching time.
+        batch_size: Record batch size, or None to let the system pick.
+        drop_last: Whether to drop the last batch if it's incomplete.
+        ensure_copy: Whether batches are always copied from the underlying base
+            blocks (not zero-copy views).
+
+    Returns:
+        An iterator over blocks of the given size that are potentially shuffled.
+    """
+    if shuffle_buffer_min_size is not None:
+        batcher = ShufflingBatcher(
+            batch_size=batch_size,
+            shuffle_buffer_min_size=shuffle_buffer_min_size,
+            shuffle_seed=shuffle_seed,
+        )
+    else:
+        batcher = Batcher(batch_size=batch_size, ensure_copy=ensure_copy)
+
+    for block in block_iter:
+        batcher.add(block)
+        while batcher.has_batch():
+            timer = stats.iter_next_batch_s.timer() if stats else nullcontext()
+            with timer:
+                batch = batcher.next_batch()
+            yield batch
+
+    # Signal to the batcher that there are no more blocks to add.
+    batcher.done_adding()
+
+    # Get any leftover batches in ShufflingBatcher.
+    while batcher.has_batch():
+        timer = stats.iter_next_batch_s.timer() if stats else nullcontext()
+        with timer:
+            batch = batcher.next_batch()
+        yield batch
+
+    # Get any remaining data.
+    if not drop_last and batcher.has_any():
+        timer = stats.iter_next_batch_s.timer() if stats else nullcontext()
+        with timer:
+            batch = batcher.next_batch()
+        yield batch
+
+
+def _format_batches(
+    block_iter: Iterator[Block], batch_format: str
+) -> Iterator[BatchType]:
+    """Given an iterator of blocks, returns an iterator of formatted batches.
+
+    Args:
+        block_iter: An iterator over blocks.
+        batch_format: The batch format to use.
+
+    Returns:
+        An iterator over formatted batches.
+    """
+    for block in block_iter:
+        batch = BlockAccessor.for_block(block).to_batch_format(batch_format)
```

Comment:
Nice simplification!

Also, we'll need the batch formatting timer here as well.
```suggestion
        timer = stats.iter_next_batch_s.timer() if stats else nullcontext()
        with timer:
            batch = BlockAccessor.for_block(block).to_batch_format(batch_format)
```

Reply:
Good catch, updated!